src/core/atomic.d

```diff
@@ -263,6 +263,52 @@ in (atomicPtrIsProperlyAligned(here), "Argument `here` is not properly aligned")
     return cast(shared)core.internal.atomic.atomicExchange!ms(cast(T*)here, cast(V)exchangeWith);
 }
 
+/**
+ * Converts a shared lvalue to a non-shared lvalue.
+ *
+ * This functions allows to treat a shared lvalue as if it was thread-local.
+ * It is useful to avoid overhead of atomic operations when access to shared data
+ * is known to be within one thread (i.e. always under a lock).
+ * ---
+ * shared static int i;
+ *
+ * // i is never used outside of synchronized {} blocks...
+ *
+ * synchronized
+ * {
+ *     ++i;                 // ERROR: cannot directly modify shared lvalue
+ *
+ *     atomicOp!"+="(i, 1); // possible overhead
+ *
+ *     // Directly modify i
+ *     assumeUnshared(i) += 1;
+ *     // or:
+ *     ++assumeUnshared(i);
+ *     // or:
+ *     i.assumeUnshared += 1;
+ * }
+ * ---
+ * Usage of this function is restricted to allowing limited lvalue access to shared instances of
+ * primitive and POD types (e.g. direct use of operators), thus it is not defined for classes.
```

Comment:
You're absolutely right. IIRC there was a `HeadUnshared` template for exactly that reason.

Reply:
I think that it would be easier to just not accept pointers (similar to what we do with classes). It seems that the intended usage is for primitive and POD types. The fundamental use case for `assumeUnshared` is to be able to modify a variable bypassing atomics - this is a straightforward case. If you have a pointer (or a class reference, for that matter) suddenly `assumeUnshared(v)` becomes ambiguous: does it apply to transitively or not? People might argue that if you use `assumeUnshared` on a pointer then it should transitively apply to all sub-members.

Given all of this, I suggest we modify this PR to not accept pointers and if the need arises we can extend the functionality to reference types also. It's better to start overly restrictive and then relax.

What do you think?

Reply:
cc @PetarKirov @MoonlightSentinel

Reply:
The intended usage for this methods is to replace atomics - hence it should not be transitive. The documentation is also pretty clear that it works only with the supplied `lvalue` which implies the pointer variable - not the target.

Rejecting pointers is fine for now but it should be enhanced to accept both pointers and classes as suggested above (the latter might need a voldemort type to allow assignments while rejecting class member access).

Reply:
@MoonlightSentinel The intended usage is to treat a shared as if it was thread-local. Been nearly 8 years (wow) now since I wrote that comment, but it stands, although it probably could stand to lose or reword the `lvalue` part. To my knowledge, `shared(T*)` is still `shared(shared(T)*)`, is it not? If so, `assumeUnshared(s).bar()` should **not** be valid, as it should return a reference to `S*`, not to a `shared(S)*`.

If you have a class, or a struct, with `shared` methods, that implies the class/struct implements necessary exclusion and/or synchronization machinery. You wouldn't be using `assumeUnshared` with instances of such classes, as that's obviated by their very design. You would, however, use it with a pointer to a regular struct or a POD, to replace the pointer value (or e.g. free and nullify it), when doing so under a lock in implementation of a `shared` method of a struct/class containing that pointer, or to call a non-`shared` method when you've already established exclusive access.

Also, there's no practical use for a mixed-bag class offering `shared` and unshared interface simultaneously, so that's kind of a contrived example.

Wrapping references to non-`shared` classes into `shared` aggregates should, IMHO, be disallowed altogether by the language (tall order, but one is entitled to some wishful thinking). Classes were designed with their own synchronization in mind, and should therefore be implemented accordingly, which is why they were excluded from this function entirely.
`shared(Klass)` is a global - there's no other use for it.

`assumeUnshared`, as I envisioned it at the time, was a hatch to escape transitivity, that is, a facility to lean upon when implementing those very `shared` methods in classes and structs, an instrument that allows to avoid having casts literally everywhere. At least, that's how I saw it before it got lost to time. What you're suggesting, OTOH, **should** be done with ugly, explicit, greppable casts.

Reply:
So what do we do here?

Reply:
What do you mean? Merge it! This goose has been on the spit for eight years, for crying out loud :)

Reply:
> @MoonlightSentinel The intended usage is to treat a shared as if it was thread-local. Been nearly 8 years (wow) now since I wrote that comment, but it stands, although it probably could stand to lose or reword the lvalue part. To my knowledge, shared(T*) is still shared(shared(T)*), is it not? If so, assumeUnshared(s).bar() should not be valid, as it should return a reference to S*, not to a shared(S)*.

That would the expected behaviour iff `assumeShared` removes `shared` from all nested layers of the supplied type. Then  the documentation be updated to reflect the behaviour w.r.t indirections and arguably classes should be accepted as well.

> assumeUnshared, as I envisioned it at the time, was a hatch to escape transitivity, that is, a facility to lean upon when implementing those very shared methods in classes and structs, an instrument that allows to avoid having casts literally everywhere. At least, that's how I saw it before it got lost to time. What you're suggesting, OTOH, should be done with ugly, explicit, greppable casts.

Guess that's a matter of personal preference. Assuming all data reached by any indirections is unshared seems more problematic to me than only removing the first layer of `shared`. But to each their own.

TL,DR: I would prefer to have consistent behaviour for struct pointers and class references. But at the very least the documentation should be changed to reflect whether `assumeUnshared` is transitive or not.